### PR TITLE
Relax NumPy requirement in UCX

### PR DESCRIPTION
**Pull request by jakirkham**
This relaxes the NumPy requirement in UCX communication. We do this by leveraging things like `struct.pack` and `struct.unpack` for metadata about the frames, which gives us simple `bytes` objects to send over the wire. Also we create a `host_array` function that will use NumPy when available, but will fallback to things like `bytearray` if not.

### CODE REVIEW

**Comment by quasiben** [comment]
Is there any expected performance improvement by using struct vs numpy ?

cc @madsbk

**Comment by jakirkham** [comment]
Mostly hoping this simpler implementation avoids red herrings while debugging issues.

**Comment by jakirkham** [comment]
In terms of performance this does improve things a little bit, but it probably doesn't impact the overall send/receive process meaningfully (though I could be wrong about that). Again this isn't really the motivation for it.

```python
In [1]: import struct                                                           

In [2]: import numpy                                                            

In [3]: l = 4 * [True, False]                                                   

In [4]: %timeit struct.pack(len(l) * "?", *l)                                   
389 ns ± 4.2 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit numpy.array(l)                                                  
838 ns ± 2.67 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

File: distributed/comm/ucx.py [hunk 1]
```diff
@@ -59,34 +60,42 @@ def init_once():
 
     ucp.init(options=ucx_config, env_takes_precedence=True)
 
+    # Find the function, `host_array()`, to use when allocating new host arrays
+    try:
```
Comment (quasiben) [inline distributed/comm/ucx.py:64]: I don't think we need the try except as ucx-py depends on numpy: https://github.com/rapidsai/ucx-py/blob/branch-0.14/setup.py#L73

Comment (jakirkham) [reply distributed/comm/ucx.py:64]: That may be so. However that would be like saying NumPy can rely on Pandas dependencies for code that Pandas uses from NumPy. In general we can't expect that to hold and it's a bit fragile. We should either make NumPy a requirement here or add a fallback. Seems safer to add this fallback here and it doesn't cost us much. Plus it should save us suffering later should things change in UCX-Py and we forget about this line here 🙂

**Comment by quasiben** [comment]
That's a good point.

**Comment by jakirkham** [comment]
Went ahead and pulled the header consolidation commit from PR ( https://github.com/rapidsai/rmm/issues/318 ) and pushed it here. That doesn't really change how the frames are serialized, but it does allow us to send all of the per frame metadata in one message.

**Comment by quasiben** [comment]
test_ucx.py passed locally with one exception:
> distributed/comm/tests/test_ucx.py::test_ucx_deserialize FAILED

though this is probably due to hostname issues.  

While I don't think this changes the overall scheme described in the notes, lines like:
> struct.pack(nframes * "?" + nframes * "Q", *cuda_frames, *sizes)

will be hard to read tomorrow.  Would you mind adding a note or two describe what packing is doing ?

File: distributed/comm/ucx.py (hunk 1 above)
Comment (quasiben) [reply distributed/comm/ucx.py:64]: i thought i hit send last night on this comment but I didn't.  It said,

That's a good point!  Thanks

**Comment by quasiben** [comment]
Thank you @jakirkham for taking the time to add those comments

**Comment by jakirkham** [comment]
Thanks Ben! 😄

Yeah I've never gotten that test to run successfully. 😞 Probably we should figure out how the config issue should be fixed and talk to OPS so it can be integrated into their provisioning scripts.

Was just about to ask if those comments seemed reasonable. Happy to adjust if needed 🙂

**Pull request merged**